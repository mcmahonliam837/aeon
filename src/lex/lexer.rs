use crate::lex::token::{Keyword, Operator, Token};
use std::{collections::VecDeque, error::Error, io::BufRead};
use utf8_chars::BufReadCharsExt;

#[derive(Debug)]
pub enum LexerError {
    IoError(std::io::Error),
}

impl std::fmt::Display for LexerError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            LexerError::IoError(err) => write!(f, "IO error: {}", err),
        }
    }
}

impl Error for LexerError {}

impl From<std::io::Error> for LexerError {
    fn from(err: std::io::Error) -> Self {
        LexerError::IoError(err)
    }
}

enum LexerState {
    InString,
    InComment,
}

enum PostProcessingCommand {
    Clear,
    ClearAndSkipPeak,
}

pub struct Lexer<R: BufRead> {
    reader: R,
}

struct LexerContext {
    current_string: String,
    tokens: Vec<Token>,
    state: VecDeque<LexerState>,
}

impl<R: BufRead> Lexer<R> {
    pub fn lex(reader: R) -> Result<Vec<Token>, LexerError> {
        let context = LexerContext {
            current_string: String::new(),
            tokens: Vec::new(),
            state: VecDeque::new(),
        };
        let mut lexer = Self { reader: reader };
        lexer.run(context)
    }

    fn run(&mut self, mut context: LexerContext) -> Result<Vec<Token>, LexerError> {
        let mut chars = self.reader.chars().peekable();

        while let Some(Ok(c)) = chars.next() {
            let peak = chars.peek().and_then(|r| r.as_ref().ok()).copied();
            let Some(cmd) = Self::process_byte(&mut context, c, peak) else {
                continue;
            };

            context.current_string.clear();
            if matches!(cmd, PostProcessingCommand::ClearAndSkipPeak) {
                chars.next();
            }
        }
        Ok(context.tokens)
    }

    fn process_byte(
        context: &mut LexerContext,
        c: char,
        peak: Option<char>,
    ) -> Option<PostProcessingCommand> {
        match context.state.back() {
            Some(LexerState::InString) => Self::process_string(context, c),
            Some(LexerState::InComment) => {
                if c == '\n' {
                    context.state.pop_back();
                }
                None
            }
            None => Self::process_normal(context, c, peak),
        }
    }

    fn process_string(context: &mut LexerContext, c: char) -> Option<PostProcessingCommand> {
        if c == '"' && !context.current_string.ends_with('\\') {
            context.state.pop_back();
            context
                .tokens
                .push(Token::LiteralString(context.current_string.clone()));
            Some(PostProcessingCommand::Clear)
        } else {
            context.current_string.push(c);
            None
        }
    }

    fn process_normal(
        context: &mut LexerContext,
        c: char,
        peak: Option<char>,
    ) -> Option<PostProcessingCommand> {
        match c {
            '\n' => {
                context.tokens.push(Token::Newline);
                Some(PostProcessingCommand::Clear)
            }
            '"' => {
                context.state.push_back(LexerState::InString);
                None
            }
            '/' if peak == Some('/') => {
                context.state.push_back(LexerState::InComment);
                Self::commit_identifier(context);
                None
            }
            c if c.is_whitespace() => {
                if !context.current_string.is_empty() {
                    Self::commit_identifier(context);
                    Some(PostProcessingCommand::Clear)
                } else {
                    None
                }
            }
            c => Self::process_operator_or_char(context, c, peak),
        }
    }

    fn process_operator_or_char(
        context: &mut LexerContext,
        c: char,
        peak: Option<char>,
    ) -> Option<PostProcessingCommand> {
        if let Ok(token) = Token::try_from(c) {
            Self::commit_and_push(context, token);
            return Some(PostProcessingCommand::Clear);
        }

        let two_char = format!("{}{}", c, peak.unwrap_or(' '));
        if let Ok(op) = Operator::try_from(two_char.as_str()) {
            Self::commit_and_push(context, Token::Operator(op));
            return Some(PostProcessingCommand::ClearAndSkipPeak);
        }

        if let Ok(op) = Operator::try_from(c.to_string().as_str()) {
            Self::commit_and_push(context, Token::Operator(op));
            return Some(PostProcessingCommand::Clear);
        }

        context.current_string.push(c);
        None
    }

    fn commit_and_push(context: &mut LexerContext, token: Token) {
        Self::commit_identifier(context);
        context.tokens.push(token);
    }

    fn commit_identifier(context: &mut LexerContext) {
        if context.current_string.is_empty() {
            return;
        }

        let token = if let Ok(keyword) = Keyword::try_from(context.current_string.as_str()) {
            Token::Keyword(keyword)
        } else {
            Token::Identifier(context.current_string.clone())
        };
        context.tokens.push(token);
        context.current_string.clear();
    }
}
