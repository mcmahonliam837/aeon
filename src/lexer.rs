use crate::token::{Keyword, Operator, Token};
use std::{collections::VecDeque, error::Error, io::BufReader};
use utf8_chars::BufReadCharsExt;

#[derive(Debug)]
pub enum LexerError {
    IoError(std::io::Error),
}

impl std::fmt::Display for LexerError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            LexerError::IoError(err) => write!(f, "IO error: {}", err),
        }
    }
}

impl Error for LexerError {}

impl From<std::io::Error> for LexerError {
    fn from(err: std::io::Error) -> Self {
        LexerError::IoError(err)
    }
}

enum LexerState {
    InString,
    InComment,
}

enum PostProcessingCommand {
    Clear,
    ClearAndSkipPeak,
}

pub struct Lexer {
    file_path: String,
    current_string: String,
    tokens: Vec<Token>,
    state: VecDeque<LexerState>,
}

impl Lexer {
    pub fn lex(file_path: &str) -> Result<Vec<Token>, LexerError> {
        let mut lexer = Lexer::new(file_path);
        lexer.run()?;
        Ok(lexer.tokens)
    }

    fn new(file_name: &str) -> Self {
        Self {
            file_path: String::from(file_name),
            current_string: String::new(),
            tokens: Vec::new(),
            state: VecDeque::new(),
        }
    }

    fn run(&mut self) -> Result<(), LexerError> {
        let input = std::fs::File::open(self.file_path.clone())?;
        let mut reader = BufReader::new(input);

        let mut position = 0;
        let mut tmp = reader.chars().peekable();
        while let Some(c) = tmp.next() {
            match c {
                Ok(c) => {
                    let peak = tmp.peek().and_then(|r| r.as_ref().ok()).copied();
                    let Some(command) = self.process_byte(c, peak) else {
                        continue;
                    };
                    match command {
                        PostProcessingCommand::Clear => {
                            self.current_string.clear();
                        }
                        PostProcessingCommand::ClearAndSkipPeak => {
                            self.current_string.clear();
                            tmp.next();
                        }
                    }
                }
                Err(ref err) if err.kind() == std::io::ErrorKind::UnexpectedEof => {
                    break;
                }
                Err(err) => {
                    return Err(LexerError::IoError(err));
                }
            }
            position += 1;
        }

        Ok(())
    }

    fn process_byte(&mut self, c: char, peak: Option<char>) -> Option<PostProcessingCommand> {
        match self.state.back() {
            None => match c {
                '\n' => {
                    self.tokens.push(Token::Newline);
                    Some(PostProcessingCommand::Clear)
                }
                '"' => {
                    self.state.push_back(LexerState::InString);
                    None
                }
                '/' if peak == Some('/') => {
                    self.state.push_back(LexerState::InComment);
                    if !self.current_string.is_empty() {
                        let token = Self::process_unknown_string(self.current_string.as_str());
                        self.tokens.push(token);
                    }
                    self.current_string.clear();
                    None
                }
                c if let Ok(token) = Token::try_from(c) => {
                    Self::commit_and_push(&mut self.tokens, token, self.current_string.as_str());
                    Some(PostProcessingCommand::Clear)
                }
                c if c.is_whitespace() && !self.current_string.is_empty() => {
                    if let Ok(keyword) = Keyword::try_from(self.current_string.as_str()) {
                        self.tokens.push(Token::Keyword(keyword));
                    } else {
                        self.tokens
                            .push(Token::Identifier(self.current_string.clone()));
                    }
                    Some(PostProcessingCommand::Clear)
                }
                c if c.is_whitespace() => None,
                c => {
                    let mut operator_chars = vec![c];
                    if let Some(peak) = peak {
                        operator_chars.push(peak);
                    };

                    let op_str = String::from_iter(operator_chars);

                    if let Ok(op) = Operator::try_from(op_str.as_str()) {
                        Self::commit_and_push(
                            &mut self.tokens,
                            Token::Operator(op),
                            self.current_string.as_str(),
                        );
                        return Some(PostProcessingCommand::ClearAndSkipPeak);
                    }

                    if let Ok(op) = Operator::try_from(c.to_string().as_str()) {
                        Self::commit_and_push(
                            &mut self.tokens,
                            Token::Operator(op),
                            self.current_string.as_str(),
                        );

                        return Some(PostProcessingCommand::ClearAndSkipPeak);
                    }

                    self.current_string.push(c);
                    None
                }
            },
            Some(LexerState::InString) => match c {
                '"' if !self.current_string.ends_with("\\") => {
                    self.state.pop_back();
                    self.tokens
                        .push(Token::LiteralString(self.current_string.clone()));
                    Some(PostProcessingCommand::Clear)
                }
                _ => {
                    self.current_string.push(c);
                    None
                }
            },
            Some(LexerState::InComment) => match c {
                '\n' => {
                    self.state.pop_back();
                    None
                }
                _ => None,
            },
        }
    }

    fn commit_and_push(tokens: &mut Vec<Token>, token: Token, str: &str) {
        if !str.is_empty() {
            tokens.push(Self::process_unknown_string(str));
        }
        tokens.push(token);
    }

    fn process_unknown_string(str: &str) -> Token {
        if let Ok(keyword) = Keyword::try_from(str) {
            Token::Keyword(keyword)
        } else {
            Token::Identifier(str.to_string())
        }
    }
}
