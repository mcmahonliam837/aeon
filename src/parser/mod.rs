pub mod ast;
pub mod block;
pub mod expression;
pub mod functions;
pub mod modules;
pub mod parser_error;
pub mod statement;
pub mod token_stream;
// pub mod variables;

use crate::{
    lex::token::Token,
    parser::{
        ast::Ast, modules::ModuleParser, parser_error::ParserError, token_stream::TokenStream,
    },
};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ParserState {
    current_module: String,
    current_function: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct ParserContext {
    stack: Vec<ParserState>,
}

impl ParserContext {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn enter_module(&mut self, module_name: String) {
        self.stack.push(ParserState {
            current_module: module_name,
            current_function: None,
        });
    }

    pub fn exit_module(&mut self) {
        self.stack.pop();
    }

    pub fn enter_function(&mut self, function_name: String) {
        self.stack.push(ParserState {
            current_module: self
                .stack
                .last()
                .expect("Cannot enter function without module")
                .current_module
                .clone(),
            current_function: Some(function_name),
        });
    }

    pub fn exit_function(&mut self) {
        self.stack.pop();
    }

    pub fn get_fully_qualified_module_name(&self) -> String {
        let mut module_name = String::new();
        let mut previous_module_name = String::new();
        for state in self.stack.iter() {
            if state.current_module == previous_module_name {
                continue;
            }
            if !module_name.is_empty() {
                module_name.push('.');
            }
            module_name.push_str(&state.current_module);
            previous_module_name = state.current_module.clone();
        }
        module_name
    }

    pub fn get_fully_qualified_function_name(&self) -> Option<String> {
        let function_name = self
            .stack
            .last()
            .and_then(|state| state.current_function.clone())?;

        Some(format!(
            "{}.{}",
            self.get_fully_qualified_module_name(),
            function_name
        ))
    }
}

pub struct Parser;

impl Parser {
    pub fn parse(tokens: &[Token]) -> Result<Ast, ParserError> {
        let mut ctx = ParserContext::new();
        let mut stream = TokenStream::new(tokens);
        let module = ModuleParser::parse(&mut ctx, &mut stream)?;
        Ok(Ast { root: Some(module) })
    }
}

#[cfg(test)]
mod token_stream_test;

#[cfg(test)]
mod parser_test;

#[cfg(test)]
mod tests {

    use super::*;

    use std::io::BufReader;

    use insta::assert_debug_snapshot;

    use crate::lex::lexer::Lexer;
    use stringreader::StringReader;

    fn parse(path: &str) -> Result<Ast, Box<dyn std::error::Error>> {
        let source = std::fs::read_to_string(path).expect("failed to load test source");
        let reader = StringReader::new(source.as_str());
        let tokens = Lexer::lex(BufReader::new(reader))?;
        Parser::parse(&tokens).map_err(Box::from)
    }

    #[test]
    fn test_hello_world() {
        assert_debug_snapshot!(parse("aeon_examples/hello_world.aeon"));
    }

    #[test]
    fn test_math() {
        assert_debug_snapshot!(parse("aeon_examples/math.aeon"));
    }

    #[test]
    fn test_modules() {
        assert_debug_snapshot!(parse("aeon_examples/modules.aeon"));
    }
}
