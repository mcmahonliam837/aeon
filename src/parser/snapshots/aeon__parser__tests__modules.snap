---
source: src/parser/mod.rs
expression: "parse(\"aeon_examples/modules.aeon\")"
---
Err(
    UnexpectedToken(
        Dot,
    ),
)
